packages/runtime/src/util/validate-plan.ts

```diff
@@ -20,6 +24,54 @@ export default (plan: ExecutionPlan) => {
   return true;
 };
 
+const assertWorkflowStructure = (plan: ExecutionPlan) => {
+  const { workflow, options } = plan;
+
+  if (!workflow || typeof workflow !== 'object') {
+    throw new ValidationError('Missing or invalid workflow key in execution plan.');
+  }
+
+  if (!Array.isArray(workflow.steps)) {
+    throw new ValidationError('The workflow.steps key must be an array.');
+  }
+
+  if (workflow.steps.length === 0) {
+    logger.warn('Warning: The workflow.steps array is empty.');
+  }
+
+  workflow.steps.forEach((step, index) => {
+    assertStepStructure(step, index);
+  });
+
+  if (options) {
+    assertOptionsStructure(options);
+  }
+};
+
+const assertStepStructure = (step: Job | Step | Trigger, index: number) => {
+  const allowedKeys = ['id', 'name', 'next', 'previous', 'adaptor', 'expression', 'state', 'configuration', 'linker'];
+
+  Object.keys(step).forEach((key) => {
+    if (!allowedKeys.includes(key)) {
+      throw new ValidationError(`Invalid key "${key}" in step ${step.id || index}.`);
+    }
+  });
+
+  if ('adaptor' in step && !('expression' in step)) {
+    throw new ValidationError(`Step ${index} with an adaptor must also have an expression.`);
+  }
+};
+
+const assertOptionsStructure = (options: WorkflowOptions) => {
+  const allowedKeys = ['timeout', 'stepTimeout', 'start', 'end', 'sanitize'];
+
+  Object.keys(options).forEach((key) => {
+    if (!allowedKeys.includes(key)) {
+      logger.warn(`Warning: Unrecognized option "${key}" in options object.`);
```

Comment:
Please don't include full stops after log messages. They're unusual and unnecessary in log output